src/states/common/base.rs

```diff
@@ -268,28 +266,50 @@ pub trait Base: Display {
     fn handle_new_deserialised_message(
         &mut self,
         src_addr: SocketAddr,
-        message: MessageWithBytes,
+        msg: MessageWithBytes,
         outbox: &mut dyn EventBox,
-    ) -> Result<Transition, RoutingError> {
-        match message {
-            MessageWithBytes::Hop(msg) => self.handle_hop_message(msg, outbox),
-            MessageWithBytes::Direct(msg, _) => {
-                let (msg, public_id) = msg.open()?;
-                let connection_info =
-                    if let Some(connection_info) = self.peer_map().get_connection_info(&src_addr) {
-                        connection_info.clone()
-                    } else {
-                        trace!(
-                            "{} - Received direct message from unconnected peer {}: {:?}",
-                            self,
-                            public_id,
-                            msg
-                        );
-                        return Ok(Transition::Stay);
-                    };
+    ) -> Result<Transition> {
+        if !self.filter_incoming_message(&msg) {
+            trace!(
+                "{} Known message: {:?} - not handling further",
+                self,
+                msg.full_crypto_hash()
+            );
 
-                self.handle_direct_message(msg, P2pNode::new(public_id, connection_info), outbox)
-            }
+            return Ok(Transition::Stay);
+        }
+
+        if self.should_relay_message(msg.message_dst()) {
+            // Relay closer to the destination or broadcast to the rest of our section.
+            self.relay_message(msg.clone())?;
```

Comment:
Why is that cloned?

Reply:
We still use the message below. Or are you saying we should pass it by reference and let the implementations of `relay_message` clone it only when necessary? That would probably be better.

Reply:
> we should pass it by reference and let the implementations of relay_message clone it only when necessary

Yes. relay_message should not need to clone generally.